design-proposal.md

```diff
@@ -0,0 +1,66 @@
+# Proposal: Design of the rkt + Kubernetes CRI
+
+## Abstract
+
+This proposal can be viewed as more a discussion than an actual proposal, because we still haven't reached a consensus on the architecture.
+
+## Backgroud
+
+As container runtime interface (CRI) proposals and original stub codes are already merged into Kubernetes, the actual work has already begun.
+All existing and future container runtimes are expected to conform the interface and provides feature parities.
+
+## Design(s)
+
+There are three different designs:
+
+- `Kubelet` --_gPRC calls_--> `rktlet service` --_invoke 'rkt' binary, or use some rkt library_--> result
+
+- `Kubelet` --_in process invoke_--> `rktshim` --_gPRC calls_--> `rkt API service` --> result
+
+- `Kubelet` --_gPRC calls_--> `rktlet service` --_gPRC calls_--> `rkt API service` --> result
+
+1. The major difference of the first two approaches is the 1st approach doesn't require any rkt code to live in the kubernetes codebase, which improves the Kubelet's modularity as a whole.
+However, it appears the rktlet code will live outside of rkt project as well since the gRPC interfaces are Kubernetes specific.
+This means for the 1st approach, we protentially have more work to do, and many of them are duplicated effort with the current rkt api service.
+Also I don't think fork/exec a rkt process is the best ideal state for the rktnetes integration, so I would expect we export a lot of rkt functionalities as a library.
```

Comment:
I'd vote for 1 since it lets rkt's CRI shim lives outside of kubelet, while not adding an extra hop. Although from the description, this option seems to demand the most work?